meilisearch-error/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+    MissingHeader,
+    NotFound,
+    RetrieveDocument,
+    SearchDocuments,
+    PayloadTooLarge,
+    UnsupportedMediaType,
+    Other,
+}
+
+impl Code {
+
+    /// ascociate a `Code` variant to the actual ErrCode
+    fn err_code(&self) -> ErrCode {
+        use Code::*;
+
+        match self {
+            // index related errors
+            CreateIndex => ErrCode::invalid("create_index", StatusCode::BAD_REQUEST),
+            IndexAlreadyExists => ErrCode::invalid("existing_index", StatusCode::BAD_REQUEST),
+            IndexNotFound => ErrCode::invalid("index_not_found", StatusCode::NOT_FOUND), InvalidIndexUid => ErrCode::invalid("invalid_index_uid", StatusCode::BAD_REQUEST),
+            OpenIndex => ErrCode::internal("open_index", StatusCode::INTERNAL_SERVER_ERROR),
+
+            // invalid state error
+            InvalidState => ErrCode::internal("invalid_state", StatusCode::INTERNAL_SERVER_ERROR),
+            // FIXME probably not an internal statuscode there
+            MissingPrimaryKey => ErrCode::internal("missing_primary_key", StatusCode::INTERNAL_SERVER_ERROR),
+            PrimaryKeyAlreadyPresent => ErrCode::internal("primary_key_already_present", StatusCode::INTERNAL_SERVER_ERROR),
+
+            // invalid document
+            MissingDocumentId => ErrCode::invalid("missing_document_id", StatusCode::BAD_REQUEST),
+            MaxFieldsLimitExceeded => ErrCode::invalid("max_field_limit_exceeded", StatusCode::BAD_REQUEST),
+
+            Filter => ErrCode::invalid("fitler", StatusCode::BAD_REQUEST),
+            Facet => ErrCode::invalid("facet", StatusCode::BAD_REQUEST),
+
+            BadParameter => ErrCode::invalid("bad_parameter", StatusCode::BAD_REQUEST),
+            BadRequest => ErrCode::invalid("bad_request", StatusCode::BAD_REQUEST),
+            RetrieveDocument => ErrCode::invalid("retrieve_document", StatusCode::BAD_REQUEST),
+            SearchDocuments => ErrCode::invalid("search_document", StatusCode::BAD_REQUEST),
+            DocumentNotFound => ErrCode::invalid("document_not_found", StatusCode::NOT_FOUND),
+            NotFound => ErrCode::invalid("not_found", StatusCode::NOT_FOUND),
+            InvalidToken => ErrCode::authentication("invalid_token", StatusCode::UNAUTHORIZED),
+            MissingHeader => ErrCode::authentication("missing_header", StatusCode::UNAUTHORIZED),
+            MissingAuthorizationHeader => ErrCode::authentication("missing_authorization_header", StatusCode::FORBIDDEN),
+            Internal => ErrCode::internal("internal", StatusCode::INTERNAL_SERVER_ERROR),
+            Maintenance =>  ErrCode::invalid("maintenance", StatusCode::SERVICE_UNAVAILABLE),
+            PayloadTooLarge => ErrCode::invalid("payload_too_large", StatusCode::PAYLOAD_TOO_LARGE),
+            UnsupportedMediaType => ErrCode::invalid("unsupported_media_type", StatusCode::UNSUPPORTED_MEDIA_TYPE),
+            _ => ErrCode::invalid("other", StatusCode::BAD_REQUEST),
+        }
```

Comment:
@curquiza @bidoubiwa @Kerollmops @eskombro This is what I came up with for error codes, WDYT?

Reply:
Some may look redundant, but I need them to report the http status

Reply:
oh and ignore "other" it's not a thing, just handy for developement

Reply:
aha `fitler`, you should probably propose a change to the english dictionnary now :)

Reply:
`SearchDocuments` and `search_document`, is there a missing `s` at the end?

Reply:
If we only talk about categories, therefore, it seems ok to me!

Reply:
Looks good 🎉 

Now, my tiny comments/suggestions:

1. `existing_index` => `index_already_exists`

Any reason the error code is named differently than your variable only here? You named it `existing_index` and the actual name of the variable seems more clear `index_already_exists`

2. `403` and `401`:

Would clarify a little bit more the difference in the naming, as `missing_header` and `missing_authorization_header` seem clear when you see both at the same time, but user will just receive one. I think `missing_header` is just not clear enough.

3. `payload_too_large` => `payload_limit_exceeded`

4. `fitler` and `facet`

You know, those are not even error names :)

Reply:
> aha `fitler`, you should probably propose a change to the english dictionnary now :)

muscle memory :/ and I just can't pick them up haha

Reply:
> fitler, facet and search_documents

yeah I know, maybe invalid_* would be better? 

>Would clarify a little bit more the difference in the naming, as missing_header and missing_authorization_header seem clear when you see both at the same time, but user will just receive one. I think missing_header is just not clear enough.

He will also receive a message, and a link to explain the error code :), the codes are just mnemonics whose purpose is to link to more thorough explaination

Reply:
> yeah I know, maybe invalid_* would be better?

Sure!

> He will also receive a message, and a link to explain the error code :), the codes are just mnemonics whose purpose is to link to more thorough explaination

You are right, I tend to think that the error code must be almost self explanatory, but if it includes several different missin header errors, then it should be fine!

Reply:
```
     Facet => ErrCode::invalid("facet", StatusCode::BAD_REQUEST),
```
Are they going to have more precise errors like primaryKey or is this in a second time?
ex: 
- `bad facet format`
- `Attribute is not a facet` (if its not in attributesForFacetting)

Reply:
yes, this is just the error code, it is given along the message that we already have

Reply:
Oke! Nice. `invalid_facets`, `invalid_filter`, `invalid_search` seems good to me.